### PR TITLE
Clear notifications for viewed chat

### LINKED ISSUE #803: Notifications don’t clear after viewing the related chat
Notifications remain visible even after opening the chat that triggered them.
They should automatically disappear once the user views that chat.

### PR DESCRIPTION
**Pull request by untreu2**
## Description
- Added automatic notification clearing when entering a chat
- Only clears notifications for the specific chat being opened
- Other chat notifications remain unaffected

### How It works
- User opens a chat
- System finds all notifications for that chat
- Those notifications are automatically removed
- User no longer sees notifications for viewed chat

### Changes
- NotificationService: Added cancelNotificationsByGroup() method
- ChatScreen: Clear notifications in initState()
- Tests: Added unit tests for notification clearing logic

https://github.com/user-attachments/assets/38d2ad1f-bcf2-4177-9ce5-d0ae9c0c3d40

## Type of Change

- [x] ✨ New feature (non-breaking change which adds functionality)
- [ ] 🛠️ Bug fix (non-breaking change which fixes an issue)
- [ ] ❌ Breaking change (fix or feature that would cause existing functionality to change)
- [ ] 🧹 Code refactor
- [ ] ✅ Build configuration change
- [ ] 📝 Documentation
- [ ] 🗑️ Chore
- [x] 🧪 Tests

## Checklist

- [x] Run `just precommit` to ensure that formatting and linting are correct
- [x] Run `just check-flutter-coverage` to ensure that flutter coverage rules are passing
- [x] Updated the `CHANGELOG.md` file with your changes (if they affect the user experience)

Fixes #803

## Summary by CodeRabbit

* **New Features**
  * Group-based notification management: notifications can be filtered and canceled by group; entering a chat automatically clears related notifications.
  * Invite notifications now include group identifiers so they join the same clearing behavior.
  * Improved notification show/cancel handling and more reliable clearing when viewing chats.

* **Tests**
  * Added comprehensive unit tests and mocks covering payload parsing, group matching, cancellation behavior, and edge cases.

* **Documentation**
  * Updated changelog to document automatic notification clearing.

### CODE REVIEW

File: test/domain/services/notification_service_test.dart [hunk 1]
```diff
@@ -0,0 +1,234 @@
+import 'dart:convert';
```
[the remark this answers is not in the document]
Comment (untreu2) [reply test/domain/services/notification_service_test.dart]: No, you're right - It was testing only that JSON parsing for notifications, I made it like that because flutter_local_notifications needs platform-spesific setup and I wrote the test only for parsing logic. Now I imported the actual service and flutter_local_notifications pkg. Nice catch, thanks.

File: test/domain/services/notification_service_test.dart [hunk 2]
```diff
@@ -0,0 +1,279 @@
...
+        }
+
+        // Verify no notifications were cancelled
+        expect(notificationsToCancel, isEmpty);
+        verify(mockPlugin.getActiveNotifications()).called(1);
+        verifyNever(mockPlugin.cancel(any));
+      });
+    });
+
+    group('actual service calls', () {
```
Comment (josefinalliende) [inline test/domain/services/notification_service_test.dart]: In my opinion, this are the only group tests that should remain, but testing that something just "returns normally" is not a useful test. The tests  in the group above are not calling the method.

File: test/domain/services/notification_service_test.dart [hunk 3]
```diff
@@ -0,0 +1,279 @@
...
+
+      test('other service methods should handle uninitialized state', () async {
+        expect(() async {
+          await NotificationService.cancelNotification(-1);
+          await NotificationService.cancelAllNotifications();
+        }, returnsNormally);
+      });
+    });
+
+    group('payload structure validation', () {
```
Comment (josefinalliende) [inline test/domain/services/notification_service_test.dart]: Please remove this tests, which is just testing jsonEncode, not the notification service class

**Comment by untreu2** [comment]
Okay, I refactored the NotificationService using parameter injection because FlutterLocalNotificationsPlugin calls native platform code (Android/iOS for now) - as I said so that doesn't work in test environment - since it was static as a dependency, I couldn't mock it and test the actual notification logic, so I added optional plugin parameters to each method allowing me to inject MockFlutterLocalNotificationsPlugin in the test to bypass the native dependency and test the service funcs. I hope I did it right.

**Review by josefinalliende — COMMENTED** [review]
Thanks for making the changes in tests. Yes, I feel dependency injection is the right approach to test that the cancel method of the notifications plugin was called in the expected scenarios 👍🏻 💯  ✅ 

Related to the fix, I just can't replicate the fix in my iPhone, I've been testing again with two phones (in this last commit: 87e21c38a832dbf5f4cfb4c4800550829be1e745 for the phone that receives the notifications, phoneA)

- Phone A has app closed
- Phone B sends and invite and loots of messages
- Phone A opens the app, accepts the invite and goes to the group screen
- Phone A in the group screen still receives all notifications for the messages B send earlier

Also, after accepting the invite happened the same
- Phone A closes the app
- Phone B sends a lot of messages
- Phone A reopens the app and goes to the group with B 
- Still receives loots of notifications for the messages that are already seen

**Comment by untreu2** [comment]
Uh yes, I think I should adjust background service too. I only tested when app was not force-quited.

**Comment by untreu2** [comment]
So the flow should be like this:

- Receive notificationB for chatA.
- Go to the chatA.
- The notification that you received for chatA (notificationB) should be cleared.
